pkg/tools/builtin/shell.go

```diff
@@ -91,27 +111,139 @@ func (h *shellHandler) RunShell(ctx context.Context, toolCall tools.ToolCall) (*
 		h.mu.Unlock()
 	}()
 
-	// Wait for the command to complete and get the result
-	err := cmd.Wait()
+	// Capture output in real-time
+	var outBuf, errBuf bytes.Buffer
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	// Read stdout
+	go func() {
+		defer wg.Done()
+		_, _ = io.Copy(&outBuf, stdoutPipe)
+	}()
 
-	// Combine stdout and stderr
-	output := outBuf.String() + errBuf.String()
+	// Read stderr
+	go func() {
+		defer wg.Done()
+		_, _ = io.Copy(&errBuf, stderrPipe)
+	}()
 
-	if err != nil {
+	// Wait for command completion with timeout
+	const quickCommandTimeout = 30 * time.Second
```

Comment:
Let's put this const at the top of the file, also 30 seconds is maybe too long? Let's try with 10 maybe?